src/Microsoft.AspNetCore.HostFiltering/HostFilteringMiddleware.cs

```diff
@@ -0,0 +1,175 @@
+        /// <param name="logger"></param>
+        /// <param name="options"></param>
+        public HostFilteringMiddleware(RequestDelegate next, ILogger<HostFilteringMiddleware> logger, 
+            IOptions<HostFilteringOptions> options, IServerAddressesFeature serverAddresses)
+        {
+            _next = next ?? throw new ArgumentNullException(nameof(next));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _options = options?.Value ?? throw new ArgumentNullException(nameof(options));
+            _serverAddresses = serverAddresses ?? throw new ArgumentNullException(nameof(serverAddresses));
+        }
+
+        /// <summary>
+        /// Runs the filtering
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        public Task Invoke(HttpContext context)
+        {
+            EnsureConfigured();
+
+            if (!CheckHost(context))
+            {
+                context.Response.StatusCode = 400;
+                if (_options.IncludedFailureMessage)
+                {
+                    context.Response.ContentLength = DefaultResponse.Length;
+                    return context.Response.Body.WriteAsync(DefaultResponse, 0, DefaultResponse.Length);
+                }
+                return Task.CompletedTask;
+            }
+
+            return _next(context);
+        }
+
+        private void EnsureConfigured()
+        {
+            if (_allowAnyNonEmptyHost == true || _allowedHosts?.Count > 0)
+            {
+                return;
+            }
+
+            var allowedHosts = new List<StringSegment>();
+            if (_options.AllowedHosts?.Count > 0)
+            {
+                if (!TryProcessHosts(_options.AllowedHosts, allowedHosts))
+                {
+                    _logger.LogDebug("Wildcard detected, all requests with hosts will be allowed.");
+                    _allowAnyNonEmptyHost = true;
+                    return;
+                }
+            }
+            else
+            {
+                if (!TryProcessHosts(_serverAddresses.Addresses.Select(address => BindingAddress.Parse(address).Host), allowedHosts))
+                {
+                    _logger.LogDebug("Wildcard detected, all requests with hosts will be allowed.");
+                    _allowAnyNonEmptyHost = true;
+                    return;
+                }
+            }
+
+            if (allowedHosts.Count == 0)
+            {
+                throw new InvalidOperationException("No allowed hosts were configured and none could be discovered from the server.");
+            }
+
+            _logger.LogDebug("Allowed hosts: " + string.Join("; ", allowedHosts));
+            _allowedHosts = allowedHosts;
+        }
+
+        // returns false if any wildcards were found
+        private bool TryProcessHosts(IEnumerable<string> incoming, IList<StringSegment> results)
+        {
+            foreach (var entry in incoming)
```

Comment:
Nit: This method is doing a decent amount. I would separate this method into two methods: one that checks if any of the strings are wildcards, and another that transforms the entries to HostString.ToUriComponent.